submission/distributor.go

```diff
@@ -147,22 +161,74 @@ func (d *Distributor) refreshRoots(ctx context.Context) map[string]error {
 	defer d.mu.Unlock()
 
 	d.logRoots = freshRoots
+	d.rootDataFull = len(d.logRoots) == len(d.logClients)
+	// Merge individual root-pools into a unified one
+	d.rootPool = ctfe.NewPEMCertPool()
+	for _, pool := range d.logRoots {
+		for _, c := range pool.RawCertificates() {
+			d.rootPool.AddCert(c)
+		}
+	}
 
 	return errors
 }
 
+// IsRootDataFull returns whether each registered Log has root-info.
+func (d *Distributor) isRootDataFull() bool {
+	d.mu.RLock()
+	defer d.mu.RUnlock()
+	return d.rootDataFull
+}
+
 // SubmitToLog implements Submitter interface.
 func (d *Distributor) SubmitToLog(ctx context.Context, logURL string, chain []ct.ASN1Cert) (*ct.SignedCertificateTimestamp, error) {
-	// TODO(Mercurrent) add implementation.
-	return nil, nil
+	lc, ok := d.logClients[logURL]
+	if !ok {
+		return nil, fmt.Errorf("No client registered for Log with URL %q", logURL)
+	}
+	return lc.AddPreChain(ctx, chain)
 }
 
 // AddPreChain runs add-pre-chain calls across subset of logs according to
 // Distributor's policy. May emit both SCTs array and error when SCTs
 // collected do not satisfy the policy.
 func (d *Distributor) AddPreChain(ctx context.Context, rawChain [][]byte) ([]*AssignedSCT, error) {
-	// TODO(Mercurrent) add implementation.
-	return []*AssignedSCT{}, nil
+	if len(rawChain) == 0 {
+		return nil, fmt.Errorf("Distributor unable to process empty chain")
+	}
+
+	// Validate and root the chain.
+	d.mu.RLock()
+	vOpts := ctfe.NewCertValidationOpts(d.rootPool, time.Time{}, false, false, nil, nil, false, nil)
+	rootedChain, err := ctfe.ValidateChain(rawChain, vOpts)
+
+	if err != nil && d.isRootDataFull() {
+		d.mu.RUnlock()
+		return nil, fmt.Errorf("Distributor unable to process cert-chain: %v", err)
+	}
+	compatibleLogs := d.ll.Compatible(rootedChain, d.logRoots)
+	d.mu.RUnlock()
+
+	// Set-up policy structs.
+	var cert *x509.Certificate
+	if len(rootedChain) != 0 {
+		cert = rootedChain[0]
+	} else {
+		cert, err = x509.ParseCertificate(rawChain[0])
+		if x509.IsFatal(err) {
+			return nil, fmt.Errorf("Distributor unable to parse cert-chainL %v", err)
+		}
+	}
+
+	groups, err := d.policy.LogsByGroup(cert, &compatibleLogs)
+	if err != nil {
+		fmt.Printf("%s\n", err)
```

Comment:
That's an interesting logical point.
Before submission start it's known the outcome won't be able to satisfy policy. (Not enough Logs to cover)
Should we emit nil-result and error or partial-result and error/warning?

Reply:
Because this is a "should never happen" scenario, I think it's important to return an error and not return a partial result. I think it's generally confusing to return both a result and an error, so we should avoid that whenever possible. The partial result (a set of SCTs, but not a sufficient set) could accidentally be used by the caller, only to later realise that they aren't sufficient to fulfill all policies.